opamp/observiq/reload_funcs.go

```diff
@@ -0,0 +1,262 @@
+// Copyright  observIQ, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package observiq
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/observiq/observiq-otel-collector/internal/logging"
+	"github.com/observiq/observiq-otel-collector/opamp"
+	"go.uber.org/zap"
+	"gopkg.in/yaml.v3"
+)
+
+func managerReload(client *Client, managerConfigPath string) opamp.ReloadFunc {
+	return func(contents []byte) (bool, error) {
+		// Unmarshal config and only pull fields out that are allowed to be updated.
+		var newConfig opamp.Config
+		if err := yaml.Unmarshal(contents, &newConfig); err != nil {
+			return false, fmt.Errorf("failed to validate config %s", ManagerConfigName)
+		}
+
+		// Check if the updatable fields are equal
+		// If so then exit
+		if client.currentConfig.CmpUpdatableFields(newConfig) {
+			return false, nil
+		}
+
+		// Going to do an update prep a rollback
+		rollbackFunc, cleanupFunc, err := prepRollback(managerConfigPath)
+		if err != nil {
+			return false, fmt.Errorf("failed to prep for rollback: %w", err)
+		}
+
+		defer func() {
+			// Cleanup rollback
+			if err := cleanupFunc(); err != nil {
+				client.logger.Warn("Failed to cleanup rollback file", zap.Error(err))
+			}
+		}()
+
+		//create a copies for rollback
+		rollBackCfg := client.currentConfig.Copy()
+		rollbackIdent := client.ident.Copy()
+
+		// Updatable config fields
+		client.currentConfig.AgentName = newConfig.AgentName
+		client.currentConfig.Labels = newConfig.Labels
+
+		// Update identity
+		client.ident.agentName = newConfig.AgentName
+		client.ident.labels = newConfig.Labels
+
+		// Write out new config file
+		// Marshal back into bytes
+		newContents, err := yaml.Marshal(client.currentConfig)
+		if err != nil {
+			// Rollback file
+			if rollbackErr := rollbackFunc(); rollbackErr != nil {
+				client.logger.Error("Rollback failed for manager config", zap.Error(rollbackErr))
+			}
+			client.ident = rollbackIdent
+			client.currentConfig = *rollBackCfg
+			return false, fmt.Errorf("failed to reformat manager config: %w", err)
+		}
+
+		// Save config file to disk
+		if err := updateConfigFile(ManagerConfigName, managerConfigPath, newContents); err != nil {
```

Comment:
just a thought. so we could have files on disk that have new "non updatable fields" right? if a weird config came in and "everything" was different (including agent name/labels), we'd only actually update the identity's name/labels, but the file would contain new out of sync info for the stuff that can't be changed on the fly.

this practically doesn't matter while the code continues executing, but if the program is every stopped and reran, then it WILL pick up ALL the completely new info in the managed config.yaml.

i have no idea if this is even a problem, just noticing it for the first time and making sure it is/isn't a problem

Reply:
That's a good call. It totally negates the security concern.

Reply:
yep, i gotcha. this won't be an issue because the collector only runs a manager config file on startup through a flag.